Guides/FirstNonNil.md

```diff
@@ -0,0 +1,43 @@
+# FirstNonNil
+
+[[Source](https://github.com/apple/swift-algorithms/blob/main/Sources/Algorithms/FirstNonNil.swift) | 
+ [Tests](https://github.com/apple/swift-algorithms/blob/main/Tests/SwiftAlgorithmsTests/FirstNonNilTests.swift)]
+
+Retrieves the first `.some` encountered while applying the given transform.
+
+This operation is available through the `firstNonNil(_:)` method on any sequence.
+
+```swift
+let value = ["A", "B", "10"].firstNonNil { Int($0) }
+// value == .some(10)
+// 
+let noValue = ["A", "B", "C"].firstNonNil { Int($0) }
+// noValue == .none
+```
+
+
+This method is analogous to `first(where:)` in how it only consumes values until
+a `.some` is found, unlike using lazy operators, which will load any sequence into a collection
+before evaluating its transforms lazily.
+
+## Detailed Design
+
+The `firstNonNil(_:)` method is added as an extension method on the `Sequence`
+protocol:
+
+```swift
+public extension Sequence {
+    func firstNonNil<Result>(_ transform: (Element) throws -> Result?)
+    rethrows -> Result? 
+}
+
+```
+
+### Naming
+
+This method’s name was selected for its comprehensibility. 
```

Comment:
I don’t think this sentence adds much to the reader. If this section isn’t mandatory, then it can be deleted. If it _is_ mandatory, then you’ll need to explain _why_ this name has good “comprehensibility.” Perhaps @natecook1000 can help you to explain since he made the choice.

Reply:
It would be great to capture the other names that came up in the issue discussion here. Why  was `firstNonNil` chosen over those?